src/software/jetson_nano/services/motor.cpp

```diff
@@ -0,0 +1,99 @@
+                           const WheelConstants_t& wheel_constants)
+
+{
+    robot_constants_ = robot_constants;
+    wheel_constants_ = wheel_constants;
+
+    int ret = 0;
+
+#define OPEN_SPI_FILE_DESCRIPTOR(motor_name, chip_select)                                \
+                                                                                         \
+    motor_name##_motor_spi_fd = open(SPI_PATHS[chip_select], O_RDWR);                    \
+    if (motor_name##_motor_spi_fd < 0)                                                   \
+    {                                                                                    \
+        LOG(FATAL) << "can't open device: " << #motor_name;                              \
+    }                                                                                    \
+                                                                                         \
+    ret = ioctl(motor_name##_motor_spi_fd, SPI_IOC_WR_MODE32, &SPI_MODE);                \
+    if (ret == -1)                                                                       \
+    {                                                                                    \
+        LOG(FATAL) << "can't set spi mode for: " << #motor_name;                         \
+    }                                                                                    \
+                                                                                         \
+    ret = ioctl(motor_name##_motor_spi_fd, SPI_IOC_WR_BITS_PER_WORD, &SPI_BITS);         \
+    if (ret == -1)                                                                       \
+    {                                                                                    \
+        LOG(FATAL) << "can't set bits for: " << #motor_name;                             \
+    }                                                                                    \
+                                                                                         \
+    ret = ioctl(motor_name##_motor_spi_fd, SPI_IOC_WR_MAX_SPEED_HZ, &SPI_SPEED_HZ);      \
+    if (ret == -1)                                                                       \
+    {                                                                                    \
+        LOG(FATAL) << "can't set max speed hz for: " << #motor_name;                     \
+    }
+
+    OPEN_SPI_FILE_DESCRIPTOR(front_left, FRONT_LEFT_MOTOR_CHIP_SELECT)
+    OPEN_SPI_FILE_DESCRIPTOR(front_right, FRONT_RIGHT_MOTOR_CHIP_SELECT)
+    OPEN_SPI_FILE_DESCRIPTOR(back_left, BACK_LEFT_MOTOR_CHIP_SELECT)
+    OPEN_SPI_FILE_DESCRIPTOR(back_right, BACK_RIGHT_MOTOR_CHIP_SELECT)
+    OPEN_SPI_FILE_DESCRIPTOR(dribbler, DRIBBLER_MOTOR_CHIP_SELECT)
+}
+
+MotorService::~MotorService() {}
+
+std::unique_ptr<TbotsProto::DriveUnitStatus> MotorService::poll(
+    const TbotsProto::DirectControlPrimitive_DirectVelocityControl& local_velocity,
```

Comment:
Yeah I thought about that as well, it boils down to who is responsible for converting from local -> per-wheel velocties. Its definitely not the primitive_executors job

It could even be a static function that takes `DirectVelocityControl`, `RobotConstants_t` and spits out `DirectPerWheelControl`

But I decided to handle it here for now, thoughts on where to do it? (i think here or static function would make sense)

Reply:
Oh I see, this is fine then.